psi4/driver/procrouting/diis.py

```diff
@@ -181,33 +232,170 @@ def extrapolate(self, *args):
+
+    def ediis_gradient(self, x):
+        c = normalize_input(x)
+        ediis_linear = np.array([entry["energy"][0] for entry in self.stored_vectors])
+        dedc = ediis_linear + np.einsum("i,ij->j", c, self.ediis_quadratic)
+
+        norm_sq = (x**2).sum()
+        dcdx = np.diag(x) * norm_sq - np.einsum("i,j->ij", x ** 2, x)
+        dcdx *= 2 / norm_sq**2
+
+        return np.einsum("i,ij->j", dedc, dcdx)
+
+    def ediis_coefficients(self):
+        from scipy.optimize import minimize
+        self.ediis_populate()
+        result = minimize(self.ediis_energy, np.ones(len(self.stored_vectors)), method="BFGS",
+                jac=self.ediis_gradient, tol=1e-6, options={"maxiter": 200})
+
+        if not result.success:
+            raise Exception("EDIIS minimization failed. File a bug, and include your entire input and output files.")
+
+        return normalize_input(result.x)
+
+    def ediis_populate(self):
+        num_entries = len(self.stored_vectors)
+
+        self.ediis_quadratic = np.zeros((num_entries, num_entries))
+        for i in range(num_entries):
+            for item_num in range(len(self.template["densities"])):
+                d = self.load_quantity("densities", i, item_num)
+                for j in range(num_entries):
+                    f = self.load_quantity("target", j, item_num)
+                    self.ediis_quadratic[i][j] += d.vector_dot(f)
+
+        diag = np.diag(self.ediis_quadratic)
+        # D_i F_i + D_j F_j - D_i F_j - D_j F_i; First two terms use broadcasting tricks
+        self.ediis_quadratic = diag[:, None] + diag - self.ediis_quadratic - self.ediis_quadratic.T
+
+        self.ediis_quadratic *= -1/2
+
+        if self.closed_shell:
+            self.ediis_quadratic *= 2
+
+    def extrapolate(self, *args, Dnorm = None):
+        """ Perform extrapolation. Must be passed in the RMS error to decide how to handle hybrid algorithms. """
+
+        if {"adiis", "ediis"}.intersection(self.engines) and Dnorm is None:
+            raise ValidationError("An extrapolation engine insists you specify the error metric.")
+
+        performed = set()
+
+        if self.engines == {"diis"}:
+            coeffs = self.diis_coefficients()
+            performed.add("DIIS")
+        elif len(self.engines) == 1:
+            blend_start = core.get_option("SCF", "INITIAL_SCF_BLEND_START")
+            blend_stop = core.get_option("SCF", "INITIAL_SCF_BLEND_STOP")
+            if Dnorm <= blend_stop:
+                return performed
+            elif self.engines == {"ediis"}:
+                coeffs = self.ediis_coefficients()
+                performed.add("EDIIS")
+            elif self.engines == {"adiis"}:
+                coeffs = self.adiis_coefficients()
+                performed.add("ADIIS")
+            else:
+                raise Exception(f"DIIS engine not recognized: {self.engines[0]}.")
+        elif self.engines == {"diis", "adiis"} or self.engines == {"diis", "ediis"}:
```

Comment:
Is there a catch for setting both `adiis` and `ediis`?

Reply:
I don't understand what you mean. It's not possible to set both through Psi options, and if both engines are selected, an error message will be raised after falling through all `elif` conditions.

Reply:
OK, that clarifies it.